libafl_qemu/src/drcov.rs

```diff
@@ -0,0 +1,145 @@
+}
+
+impl QemuDrCovMetadata {
+    #[must_use]
+    pub fn new() -> Self {
+        Self { current_id: 0 }
+    }
+}
+
+libafl::impl_serdeany!(QemuDrCovMetadata);
+
+#[derive(Debug)]
+pub struct QemuDrCovHelper {
+    filter: QemuInstrumentationFilter,
+    filename: PathBuf,
+    drcov_len: usize,
+}
+
+impl QemuDrCovHelper {
+    #[must_use]
+    pub fn new(filter: QemuInstrumentationFilter, filename: PathBuf) -> Self {
+        Self {
+            filter,
+            filename,
+            drcov_len: 0,
+        }
+    }
+
+    #[must_use]
+    pub fn must_instrument(&self, addr: u64) -> bool {
+        self.filter.allowed(addr)
+    }
+}
+
+impl<S> QemuHelper<S> for QemuDrCovHelper
+where
+    S: UsesInput + HasMetadata,
+{
+    fn init_hooks<QT>(&self, hooks: &QemuHooks<'_, QT, S>)
+    where
+        QT: QemuHelperTuple<S>,
+    {
+        hooks.blocks_raw(Some(gen_unique_block_ids::<QT, S>), Some(trace_block));
+    }
+
+    fn pre_exec(&mut self, _emulator: &Emulator, _input: &S::Input) {}
+
+    fn post_exec(&mut self, emulator: &Emulator, _input: &S::Input) {
+        if DRCOV_IDS.lock().unwrap().len() > self.drcov_len {
+            println!("New DrCov lenght = {}", DRCOV_IDS.lock().unwrap().len());
+            let mut drcov_vec = Vec::<DrCovBasicBlock>::new();
+            for id in DRCOV_IDS.lock().unwrap().iter() {
+                for (pc, idm) in DRCOV_MAP.lock().unwrap().iter() {
+                    if *idm == *id {
+                        let block = pc2basicblock(*pc, emulator);
+                        let mut block_len = 0;
+                        for instr in &block {
+                            block_len += instr.insn_len;
+                        }
+                        block_len -= &block.last().unwrap().insn_len;
+                        drcov_vec
+                            .push(DrCovBasicBlock::new(*pc as usize, *pc as usize + block_len));
+                    }
+                }
+            }
+
+            let mut rangemap = RangeMap::<usize, (u16, String)>::new();
+            rangemap.insert(
+                (GuestAddr::MIN as usize)..(GuestAddr::MAX as usize),
+                (0, "test".to_string()),
+            );
+            DrCovWriter::new(&rangemap)
+                .write(&self.filename, &drcov_vec)
+                .expect("Failed to write coverage file");
+        }
+        self.drcov_len = DRCOV_IDS.lock().unwrap().len();
+    }
+}
+
+pub fn gen_unique_block_ids<QT, S>(
+    hooks: &mut QemuHooks<'_, QT, S>,
+    state: Option<&mut S>,
+    pc: GuestAddr,
+) -> Option<u64>
+where
+    S: HasMetadata,
+    S: UsesInput,
+    QT: QemuHelperTuple<S>,
+{
+    if let Some(h) = hooks.helpers().match_first_type::<QemuDrCovHelper>() {
+        if !h.must_instrument(pc.into()) {
+            return None;
+        }
+    }
+    let state = state.expect("The gen_unique_block_ids hook works only for in-process fuzzing");
+    if state.metadata().get::<QemuDrCovMetadata>().is_none() {
```

Comment:
Don't `get` and `get_mut` afterwards, this is two lookups instead of one. Instead, directly use metadata_mut and add to the map if it's none

Reply:
Or just add it before and assume it's `Some`

Reply:
Or better yet, use the the block hooks with closure that can take a Rc (or Arc) to the map